### PR TITLE
TASK-53273:add announcement services and API

### PR DESCRIPTION
**Pull request by GouadriaHanen**

### CODE REVIEW

File: extension/src/main/webapp/WEB-INF/conf/gamification/storage/rule/rules-storage-configuration.xml [hunk 1]
```diff
@@ -29,8 +29,15 @@ Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
...
-
+    <!-- Challenge storage-->
+    <component>
+        <type>org.exoplatform.addons.gamification.storage.ChallengeStorage</type>
+    </component>
+    <!-- Challenge Service Layer-->
+    <component>
+        <key>org.exoplatform.addons.gamification.service.ChallengeService</key>
+        <type>org.exoplatform.addons.gamification.service.configuration.ChallengeServiceImpl</type>
+    </component>
```
Comment (Rdidmohamedelhadi) [inline extension/src/main/webapp/WEB-INF/conf/gamification/storage/rule/rules-storage-configuration.xml]: delete this please is already added in completion-configuration.xml!!

File: pom.xml [hunk 2]
```diff
@@ -19,6 +19,12 @@ Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
     <modelVersion>4.0.0</modelVersion>
+    <dependencies>
+        <dependency>
+            <groupId>javax.ws.rs</groupId>
+            <artifactId>jsr311-api</artifactId>
+        </dependency>
+    </dependencies>
```
Comment (Rdidmohamedelhadi) [inline pom.xml]: why you needed this !!

File: services/src/main/java/org/exoplatform/addons/gamification/service/configuration/AnnouncementServiceImpl.java [hunk 3]
```diff
@@ -31,27 +35,87 @@ public AnnouncementServiceImpl(AnnouncementStorage announcementStorage,
...
+    announcement = announcementStorage.saveAnnouncement(announcement);
+    //TODO to be added after adding listener  and activity processeur
+    /*
+    try {
+      listenerService.broadcast(ANNOUNCEMENT_ACTIVITY_EVENT, this, announcement);
+
+    } catch (Exception e) {
+      LOG.error("Unexpected error", e);
+    }
+    */
```
Comment (Rdidmohamedelhadi) [inline services/src/main/java/org/exoplatform/addons/gamification/service/configuration/AnnouncementServiceImpl.java]: delete comment please !

File: services/src/main/java/org/exoplatform/addons/gamification/service/mapper/EntityMapper.java [hunk 4]
```diff
@@ -66,6 +68,9 @@ public static RuleEntity toEntity(Challenge challenge) {
     }
     challengeEntity.setAudience(challenge.getAudience());
     challengeEntity.setManagers(challenge.getManagers());
+    challengeEntity.setScore(20); // TODO: to be changed when add field score to DTO
+    challengeEntity.setArea("social"); // TODO: to be changed when add field domain to DTO
+    challengeEntity.setDomainEntity(DomainMapper.domainDTOToDomain(Utils.getDomainByTitle("social"))); // TODO: to be changed when add field domain to DTO
```
Comment (Rdidmohamedelhadi) [inline services/src/main/java/org/exoplatform/addons/gamification/service/mapper/EntityMapper.java]: check the last commit in feature/rewards please, and delete the default value of area and domain

File: services/src/main/java/org/exoplatform/addons/gamification/storage/AnnouncementStorage.java [hunk 5]
```diff
@@ -4,25 +4,58 @@
...
-  public List<Announcement> findAllAnnouncementByChallenge(Long challengeId, int offset, int limit) {
-    List<GamificationActionsHistory> announcementEntities= gamificationHistoryDAO.findAllAnnouncementByChallenge(challengeId, offset, limit);
-    return EntityMapper.fromAnnouncementEntities(announcementEntities);
-  }
+    private GamificationHistoryDAO announcementDAO;
+
+    private ChallengeStorage challengeStorage;
+
+    public static final long  MILLIS_IN_A_DAY = 1000 * 60 * 60 * 24;  // NOSONAR
+
```
Comment (Rdidmohamedelhadi) [inline services/src/main/java/org/exoplatform/addons/gamification/storage/AnnouncementStorage.java:20]: check sonar

File: services/src/main/java/org/exoplatform/addons/gamification/utils/Utils.java [hunk 6]
```diff
@@ -176,4 +177,8 @@ public static Long countAnnouncementsByChallenge(Long challengeId) {
       return 0l;
     }
   }
+   public static Long getUserGlobalScore( String earnerId){
+    return CommonsUtils.getService(GamificationService.class).computeTotalScore(earnerId);
+   }
```
Comment (Rdidmohamedelhadi) [inline services/src/main/java/org/exoplatform/addons/gamification/utils/Utils.java]: check NPE

**Review by Rdidmohamedelhadi — CHANGES_REQUESTED** [review]

File: extension/src/main/webapp/WEB-INF/conf/gamification/storage/rule/rules-storage-configuration.xml (hunk 1 above)
Comment (GouadriaHanen) [reply extension/src/main/webapp/WEB-INF/conf/gamification/storage/rule/rules-storage-configuration.xml]: just reorganized each service in his related XML file!

File: pom.xml (hunk 2 above)
Comment (GouadriaHanen) [reply pom.xml]: removed!

File: services/src/main/java/org/exoplatform/addons/gamification/service/configuration/AnnouncementServiceImpl.java (hunk 3 above)
Comment (GouadriaHanen) [reply services/src/main/java/org/exoplatform/addons/gamification/service/configuration/AnnouncementServiceImpl.java]: done!

File: services/src/main/java/org/exoplatform/addons/gamification/service/mapper/EntityMapper.java (hunk 4 above)
Comment (GouadriaHanen) [reply services/src/main/java/org/exoplatform/addons/gamification/service/mapper/EntityMapper.java]: area removed! for score and domain, they should not be null and there is no way to get them while mapping from DTO to entity! I need a default value to create the announcement!

File: services/src/main/java/org/exoplatform/addons/gamification/utils/Utils.java (hunk 6 above)
Comment (GouadriaHanen) [reply services/src/main/java/org/exoplatform/addons/gamification/utils/Utils.java]: done!

File: services/src/main/java/org/exoplatform/addons/gamification/service/mapper/GamificationActionsHistoryMapper.java [hunk 7]
```diff
@@ -11,7 +11,7 @@ private GamificationActionsHistoryMapper() {
   }
 
   public static GamificationActionsHistoryDTO fromEntity(GamificationActionsHistory gamificationActionsHistoryEntity) {
-    GamificationActionsHistoryDTO gActionsHistoryDTO = new GamificationActionsHistoryDTO(
```
Comment (Rdidmohamedelhadi) [inline services/src/main/java/org/exoplatform/addons/gamification/service/mapper/GamificationActionsHistoryMapper.java:14]: to deleted please

File: services/src/main/java/org/exoplatform/addons/gamification/service/mapper/EntityMapper.java [hunk 8]
```diff
@@ -2,6 +2,7 @@
 
 import org.apache.commons.collections.CollectionUtils;
 import org.apache.commons.lang3.StringUtils;
+import org.exoplatform.addons.gamification.IdentityType;
```
Comment (Rdidmohamedelhadi) [inline services/src/main/java/org/exoplatform/addons/gamification/service/mapper/EntityMapper.java]: delete this please

File: services/src/main/java/org/exoplatform/addons/gamification/utils/Utils.java [hunk 9]
```diff
@@ -176,4 +177,8 @@ public static Long countAnnouncementsByChallenge(Long challengeId) {
       return 0l;
     }
   }
+   public static Long getUserGlobalScore( String earnerId){
+    return StringUtils.isBlank(earnerId) ? null : CommonsUtils.getService(GamificationService.class).computeTotalScore(earnerId);
+   }
```
Comment (Rdidmohamedelhadi) [inline services/src/main/java/org/exoplatform/addons/gamification/utils/Utils.java]: change this please see example : https://github.com/exoplatform/agenda/blob/develop/agenda-services/src/main/java/org/exoplatform/agenda/notification/builder/DatePollNotificationBuilder.java#L136-L141

**Review by Rdidmohamedelhadi — CHANGES_REQUESTED** [review]
fix indentation

File: services/src/main/java/org/exoplatform/addons/gamification/rest/AnnouncementRest.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
+package org.exoplatform.addons.gamification.rest;
+
+import io.swagger.annotations.*;
+import org.apache.commons.lang3.StringUtils;
+import org.exoplatform.addons.gamification.service.AnnouncementService;
+import org.exoplatform.addons.gamification.service.dto.configuration.Announcement;
+import org.exoplatform.addons.gamification.service.dto.configuration.AnnouncementRestEntity;
```
Comment (Rdidmohamedelhadi) [inline services/src/main/java/org/exoplatform/addons/gamification/rest/AnnouncementRest.java]: delete this

File: services/src/main/java/org/exoplatform/addons/gamification/rest/AnnouncementRest.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+import org.exoplatform.common.http.HTTPStatus;
+import org.exoplatform.services.log.ExoLogger;
+import org.exoplatform.services.log.Log;
+import org.exoplatform.services.rest.resource.ResourceContainer;
+
+import javax.annotation.security.RolesAllowed;
+import javax.ws.rs.*;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.util.ArrayList;
```
Comment (Rdidmohamedelhadi) [inline services/src/main/java/org/exoplatform/addons/gamification/rest/AnnouncementRest.java]: delete this

**Review by Rdidmohamedelhadi — CHANGES_REQUESTED** [review]

**Review by Rdidmohamedelhadi — APPROVED** [review]

**Review by ahamdi — APPROVED** [review]

**Pull request merged**